include/merlin_hashtable.cuh

```diff
@@ -251,16 +251,17 @@ class HashTable {
       check_evict_strategy(metas);
     }
 
+    // Block exclusive access, unless table structure cannot change anymore.
+    std::shared_lock<std::shared_timed_mutex> lock(table_mutex_);
```

Comment:
First need to get lock, to make sure others cannot call `reserve(...)` anymore. Then, "if no further growth possible" is ensured, can release the lock prematurely.

Reply:
Arguably, the `bool` can only be set at the end before the unique_lock is released by `reserve`. However, that might be missed, unless the `bool` is also declared `mutable`. So better, do it like this, which makes access water-tight.